### PR TITLE
Add Context checks to factory caching and simplify factory caching between factories and statics.

### PR DESCRIPTION
**Pull request by jlaanstra**
Alternative implementation that maintains static classes compared to #1359

### CODE REVIEW

File: src/cswinrt/strings/WinRT.cs [hunk 1]
```diff
@@ -501,33 +547,144 @@ public unsafe ObjectReference<I> _ActivateInstance<I>()
...
+#if NET
+    internal class BaseFactory
+    {
+        private IObjectReference _factory;
+        public IObjectReference Value { get => _factory; }
+#else
+    internal class BaseFactory<I>
     {
-        public ActivationFactory() : base(typeof(T).Namespace, typeof(T).FullName) { }
+        private ObjectReference<I> _factory;
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:587]: Shouldn't this also be volatile? It's being written to concurrently with volatile writes in that `cmpxgh` below.
```suggestion
        private volatile ObjectReference<I> _factory;
```

**Review by Sergio0694 — COMMENTED** [review]
Just leaving a couple more comments 😄

File: src/cswinrt/strings/WinRT.cs [hunk 2]
```diff
@@ -448,31 +491,65 @@ public static unsafe (ObjectReference<IActivationFactoryVftbl> obj, int hr) GetA
...
             Marshal.ThrowExceptionForHR(Platform.CoDecrementMTAUsage(_mtaCookie));
         }
-    }
-
+    }
+
     internal class BaseActivationFactory
     {
-        private readonly ObjectReference<IActivationFactoryVftbl> _IActivationFactory;
+        private Tuple<ObjectReference<IActivationFactoryVftbl>, IntPtr> _IActivationFactory;
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:498]: This should still be marked as `volatile`, no? You're reading it here concurrently, but writing with an interlocked op.

File: src/cswinrt/strings/WinRT.cs [hunk 3]
```diff
@@ -502,32 +580,117 @@ public unsafe ObjectReference<I> _ActivateInstance<I>()
...
 #endif
-            I>() => _factory._ActivateInstance<I>();
+            I>() => _instance._ActivateInstance<I>();
+    }
+
+#if NET
+    internal class BaseFactory
+    {
+
+        private Tuple<IObjectReference, IntPtr> _factory;
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:611]: Same here for this, shouldn't it also be `volatile`?

File: src/cswinrt/strings/WinRT.cs [hunk 4]
```diff
@@ -502,32 +580,117 @@ public unsafe ObjectReference<I> _ActivateInstance<I>()
...
+        public static ObjectReference<I> As<I>() => _instance._As<I>();
+        public static IObjectReference As(Guid iid) => _instance._As(iid);
 
-#if NET
-        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2091:RequiresUnreferencedCode", 
-            Justification = "No members of the generic type are dynamically accessed in this code path.")]
-#endif
         public static ObjectReference<I> ActivateInstance<
 #if NET
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.None)]
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:602]: Is this actually needed? Isn't `I` already not annotated for any dynamic access in this context? 🤔

File: src/cswinrt/strings/WinRT.cs [hunk 5]
```diff
@@ -491,7 +568,8 @@ public BaseActivationFactory(string typeNamespace, string typeFullName)
         public unsafe ObjectReference<I> _ActivateInstance<I>()
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:568]: Also double checking: I see you removed the `[UnconditionalSuppressMessage]` below. Is this one still needed or can it also be removed after the other changes you did?

**Review by Sergio0694 — COMMENTED** [review]
Left a few comments 🙂
Also, Q: I see you removed `INativeGuid`, is that just because it'll all be done in another PR?

File: src/cswinrt/strings/WinRT.cs [hunk 6]
```diff
@@ -448,31 +491,65 @@ public static unsafe (ObjectReference<IActivationFactoryVftbl> obj, int hr) GetA
...
+            get
+            {
+                var existingInstance = _IActivationFactory;
+                if (existingInstance != null && (existingInstance.Item2 == IntPtr.Zero || existingInstance.Item2 == Context.GetContextToken()))
+                {
+                    return existingInstance.Item1;
+                }
+
+                var newFactory = InitializeFactory();
+                _IActivationFactory = newFactory;
```
Comment (Sergio0694) [inline src/cswinrt/strings/WinRT.cs:511]: Q: two racing threads might end up returning different instances of the activation factory. Is that a concern?

Comment (jlaanstra) [reply src/cswinrt/strings/WinRT.cs:511]: That is not a concern. Caching it is opportunistic.

File: src/cswinrt/code_writers.h [hunk 7]
```diff
@@ -1805,87 +1808,89 @@ remove => %;
...
-                cache_vftbl_type);
+            auto statics_type_name = staticsType.TypeName();
             w.write(R"(
-internal sealed class _% : ABI.%.%
-{
-public _%() : base(%()) { }
-%
-internal static % Instance => _instance;
-}
+private static % _% = new %("%.%", %.IID);
```
Comment (Sergio0694) [inline src/cswinrt/code_writers.h:1854]: Is this ever reassigned at runtime? If not, shouldn't this be `readonly` to improve codegen?

File: src/cswinrt/code_writers.h [hunk 8]
```diff
@@ -1805,87 +1808,89 @@ remove => %;
...
+    }
 
-%
-internal static % Instance => (%)_instance;
-}
+    void write_activation_factory_static_cache_definition(writer& w, TypeDef const& classType)
+    {
+        auto class_type_name = classType.TypeName();
+        w.write(R"(
+private static BaseActivationFactory _%Factory = new BaseActivationFactory("%.%");
```
Comment (Sergio0694) [inline src/cswinrt/code_writers.h:1869]: Same for this one, shouldn't it be `readonly`, or could this ever be reassigned?

**Pull request closed without merging**